### ISSUE TITLE
Execute arbitrary sql statements

### ISSUE THREAD

**Issue by mhewedy** [go-rel/rel#109]
I have an update statement that looks like:

```
update matches m
    set viewed = true
    where m.request_id in (
        select r.id
        from requests r
                    join users u on r.user_id = u.id
        where u.username = $1)
```

And I need to execute via `rel` but I couldn't.  ( and I don't want to use rel `dsl` here 😄  )

If it is a select query, I could use `SQL` function  and pass it to `FindAll` function.
I tried the `Update` function but doesn't seem to work.

I had to use the underlaying `*sql/DB` `Exec` method.

**Comment by Fs02** [comment]
> And I need to execute via rel but I couldn't. ( and I don't want to use rel dsl here 😄 )
No worries, building dsl for complex query isn't a goal, at least for now 🤣 

Insert and Update method is very specific, while the functionality you want is more general, it's closer to sql.DB Exec, therefore making Insert and Update support SQL, I think is not the cleanest way to go.

We can try adding Exec method to the Repository, but it'll be a wrapper to adapter then to sql.DB.

Do you have any suggestion or idea?

**Comment by mhewedy** [comment]
> Insert and Update method is very specific, while the functionality you want is more general, it's closer to sql.DB Exec, therefore making Insert and Update support SQL, I think is not the cleanest way to go.

100% agree on this.

> We can try adding Exec method to the Repository, but it'll be a wrapper to adapter then to sql.DB.

I think this is the best suggestion till the moment.  Not worth it to have a different interface to execute such ad-hoc statements.

**Comment by Fs02** [comment]
> I think this is the best suggestion till the moment. Not worth it to have a different interface to execute such ad-hoc statements.
Sorry, I don't think I understand this very well.
Did you mean, you are positive with adding "Exec method to the Repository", but you also thinking that it's not worth implementing?

**Comment by mhewedy** [comment]
No I would say I agree for your suggestion.

And then said I think if to impement the method under a new interface
(rathar than Repository) wouldn't worth it. (I emphasis on thr first
statement here about being agree on your suggestion)

Sorry for being not clear.

**Comment by Fs02** [comment]
Okay, then let's go with adding Exec method to the repository.

The new method will looks like this, I don't think we need to return last inserted id because not all db supports it:

```go
// Exec arbitary statements.
// returns rows affected and errors.
Exec(ctx context.Context, statement string, args ...interface{}) (int64, error)
```

This should not be hard to do, as the adapter already have this [method](https://github.com/Fs02/rel/blob/master/adapter/sql/adapter.go#L89), just need to implement `rel.Repository` and `reltest.Repository` (and docs).

**Comment by mhewedy** [comment]
I noticed you ignore error here:
https://github.com/Fs02/rel/blob/c7b126f33cb90e773956ad66ee10a4fa5cdc6660/adapter/sql/adapter.go#L98

why not pass it to the caller, what is the point?

Also, I think we don't need to hide a piece "num inserted" which is info that might need in future cases

**Comment by Fs02** [comment]
> why not pass it to the caller, what is the point?

Some database like `postgres`, doesn't support `LastInsertId`, which (I never try) may returns error when it's called, imo, instead of passing it to the caller, we should panic here, because we don't want that to be error at all - feel free to improve this 😄 

> Also, I think we don't need to hide a piece "last inserted id" which is info that might need in future cases

So far I only cross check on gorm to see if we can retrieve `LastInsertedId` and we can't. but if you strongly believe this will be useful, sure we can add that as well, just make sure the behavior is documented and expected on DB that doesn't support LastInsertedId.

**Comment by KMPolakowski** [comment]
@Fs02 pls assign to me

**Comment by Fs02** [comment]
Done, thanks!

**Comment by KMPolakowski** [comment]
@Fs02 would you mind if I change the Exec method's return value to sql.Result, error:
`func (a *Adapter) Exec(ctx context.Context, statement string, args []interface{}) (sql.Result, error) {`

That way one can elegantly retrieve the `lastInsertedId` himself and check for errors.

**Comment by Fs02** [comment]
@KMPolakowski hmm, I'm still not convinced why we need to return lastInsertedId? why normal insertion is not sufficient?
I just want to avoid bloating the API with something we think is needed, but in practice is never used, or very rarely used.

**Comment by KMPolakowski** [comment]
@Fs02 Last Insert Id is sent by mysql for a reason in it's OK_Packet. I don't really see the value of it as a user because mostly I'd only check for err and ignore the result struct altogether. Still the method signature with Result is better type information to me because otherwise I'm getting some int64 that says nothing in itself and have to check in comments, docs or in code whether this is last_insert_id or rows_affected. Altogether I'm fine with both variants as this is very minor.

**Comment by Fs02** [comment]
@KMPolakowski hmmm, I prefer to return only `affectedRows int64`, if it's only one int, shouldn't be very hard to reason what number is that.

I would like to avoid returning `sql.Result`, as it will make it harder to implement adapter that doesn't rely on `database/sql`.

**Comment by KMPolakowski** [comment]
@Fs02 fine for me

**Comment by Fs02** [comment]
released on v0.14.0, thanks all 👍 👍
